### ISSUE TITLE
conan install sporadically requires authentication

### ISSUE THREAD

**Issue by jakuborava** [conan-io/conan#6262]
Hello, Conan devs,

first of all, thank you all for your hard work on this great tool! I really appreciate it.

Now to the interesting bug. Some developers in my company started to use Conan, in most cases, there are no issues (again, good work! 😃). Unfortunately, one developer today had a problem to run `conan install` command.

What happened:
After running `conan install` all the conanfile.py files are downloaded and dependencies are resolved. So far it is ok. Then, let's say, 5 packages are downloaded (this happens randomly, sometimes more, sometimes less) and when the sixth one is being downloaded authentication is required. The package comes from the same repository as the previous packages. When the developer tried to re-run `conan install`, he got in the log "Error, the file to download already exists" with path to the conanmanifest.txt of the package which download failed. Then he removed the whole package directory from conan cache and with next `conan install` the problematic package was downloaded successfully, but another package failed and then this story repeats until all packages are downloaded.

In artifactory.log I found:
2019-12-19 13:44:47,948 [http-nio-8081-exec-2] [ERROR] (o.a.r.c.e.m.GlobalExceptionMapper:48) - 0
java.lang.ArrayIndexOutOfBoundsException: 0 

This is the first time when something like this happened. Other developers are running with an anonymous user without a problem. I also tried to reproduce the bug on a laptop with a configuration I specified below and an anonymous user, but no luck.

I can provide you an obfuscated logs (Conan, Artifactory), but even they are obfuscated I would rather send it via email if that is ok.

### Environment Details
  * Operating System+version: Windows 7 64bit
  * Compiler+version: MSVC 9
  * Conan version: 1.21.0
  * Python version: 3.8.0
  * Artifactory version: 6.10.1

**Comment by memsharded** [comment]
Hi @jakuborava 

Thanks for your kind words! :) 

Yes, this seems the typical bug/issue very challenging to debug, because it is very challenging to reproduce it. Sure, a first look to the Conan and Artifactory logs might be useful (try to activate the CONAN_TRACE_FILE to get a trace of the API calls to the server, besides the terminal output, which you could please also activate the debug level for the logs). 

Sure, if you prefer to send it in private, you can email info@conan.io. Thanks!

**Comment by jakuborava** [comment]
Hello @memsharded

Thanks for the quick reply.

Right, I am aware that kind of bugs can be challenging. I got from developer conan logs with CONAN_TRACE_FILE activated and log level was set to debug. Unfortunately, I forgot to get a terminal output from the developer. What happened in the time when the problem occurred in the logs, was that he had in terminal prompt for authentication.

Logs were sent. Thank you for looking at it.

**Comment by memsharded** [comment]
Hi @jakuborava 

We have been having a look at the logs, seems difficult to know, a couple of quick questions:

If this user, in that same machine "logs out" and becomes anonymous (with ``conan user --clean``), everything works? Could you please ask him to try that? Then log-in again and repeat?

If other users try to do the same install, but making sure they are logged-in authenticated in Artifactory, it fails?

Thanks for the feedback!

**Comment by memsharded** [comment]
This trace in Artifactory is still interesting:

```bash
2019-12-19 13:44:47,948 [http-nio-8081-exec-2] [ERROR] (o.a.r.c.e.m.GlobalExceptionMapper:48) - 0
java.lang.ArrayIndexOutOfBoundsException: 0
	at org.artifactory.addon.conan.rest.ConanUsersHandlerImpl.extractUser(ConanUsersHandlerImpl.java:93)
	at org.artifactory.addon.conan.rest.ConanUsersHandlerImpl.authenticate(ConanUsersHandlerImpl.java:61)
	at org.jfrog.repomd.conan.rest.users.ConanUsersSubResource.lambda$authenticate$1(ConanUsersSubResource.java:35)
	at org.jfrog.repomd.conan.rest.ConanVersionHeadersResponseDecorator.handle(ConanVersionHeadersResponseDecorator.java:53)
	at org.jfrog.repomd.conan.rest.users.ConanUsersSubResource.authenticate(ConanUsersSubResource.java:35)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

Please @jgsogo if you could inspect the code around that to investigate if there some possible bug there, that could help.

**Comment by jakuborava** [comment]
Hello @memsharded,

thank you for the feedback. To answer your questions:

> If this user, in that same machine "logs out" and becomes anonymous (with conan user --clean), everything works? Could you please ask him to try that? Then log-in again and repeat?

We actually tried that so I can answer right away. The result was the same - some packages were downloaded and then authentication was needed.  When he authenticated it was accepted without problem but then during downloading package, he got "ERROR: Exception: <class 'conans.errors.ForbiddenException'> Permission denied for user: " in a terminal. For this I do not have logs from Conan or Artifactory. I will be able to collect them after the holidays (i.e. in 2-3 weeks) if needed.

> If other users try to do the same install, but making sure they are logged-in authenticated in Artifactory, it fails?

No, I have tested that now on Win7 and Win10 with a read-only user logged-in and everything works fine. I have created this user manually in Artifactory, the developer's account is from LDAP, but this probably does not matter.

**Comment by jgsogo** [comment]
Hi! Having a look at the source I've found the place where it is failing (I cannot share the snippet here).

It looks like Artifactory is receiving an authentication header like `"Authorization: Basic "`. In the logs, all those headers are obfuscated, maybe you can investigate in this direction and have a look at the `Authorization` headers that are being sent to Artifactory. Is it possible?

**Comment by memsharded** [comment]
> the developer's account is from LDAP, but this probably does not matter.

Do you mean that these sporadic errors are for an account that uses LDAP? Maybe it is related? I wouldn't discard any idea quickly, probably worth investigating. 

Yes, we understand these might be holidays, looking forward the logs in 2-3 weeks. Cheers!

**Comment by jakuborava** [comment]
Hi guys,
sorry for the late response, I had a lot on my plate. Unfortunately, I have bad news because the developer who had this issue reinstalled his laptop to Win10 before I was able to collect the logs or do tests. After the reinstallation problem disappeared. So I will close the issue and if the problem will occur on another machine I will reopen it and I will try to collect all the stuff we discussed here. Thank you a lot for your help!
